core/site.go

```diff
@@ -409,23 +415,39 @@ func (site *Site) updateMeters() error {
 				site.log.ERROR.Printf("battery %d power: %v", i+1, err)
 			}
 
+			var capacity float64
 			soc, err := meter.(api.Battery).Soc()
 
 			if err == nil {
-				site.batterySoc += soc
+				// weigh soc by capacity and accumulate total capacity
+				weighedSoc := soc
+				if m, ok := meter.(api.BatteryCapacity); ok {
+					capacity = m.Capacity()
+					totalCapacity += capacity
+					weighedSoc *= capacity
+				}
+
+				site.batterySoc += weighedSoc
 				site.log.DEBUG.Printf("battery %d soc: %.0f%%", i+1, soc)
 			} else {
-				err = fmt.Errorf("battery %d soc: %v", i+1, err)
-				site.log.ERROR.Println(err)
+				site.log.ERROR.Printf("battery %d soc: %v", i+1, err)
 			}
 
 			mm[i] = batteryMeasurement{
-				Power: power,
-				Soc:   soc,
+				Power:    power,
+				Soc:      soc,
+				Capacity: capacity,
 			}
 		}
 
-		site.batterySoc /= float64(len(site.batteryMeters))
+		site.publish("batteryCapacity", math.Round(totalCapacity))
+
+		// convert weighed socs to total soc
+		if totalCapacity == 0 {
+			totalCapacity = float64(len(site.batteryMeters))
```

Comment:
Wie bereits diskutiert, ist diese Rechnung dann falsch, wenn manche Batterien eine Kapazität haben und andere nicht.

Ein Beispiel für eine sinnvolle Anwendung für Batterien ohne Kapazität:
_3 Batterien, 2 mit 100Ah, eine mit 200Ah, alle bei 75% soc. Ich gebe nur bei der großen, eine Kapazität von 2 an, bei den beiden anderen nichts, da ich die unverändert lassen möchte. Für die Berechnung des Geasmt-SOC ist ja nur die Relation der Batteriegrößen zueinander wichtig._

Hier wird in dem Beispiel nun mit einer Gesamtkapazität von 2 gerechnet, es sind tatsächlich aber 4. Das Ergebnis ist ein Gesamt-SOC von 150%. Der ursprüngliche Algorithmuis hätte korrekte 75% ergeben.

Ich bin der Meinung, dass keine akzeptierte Konfiguration zu solch einem unsinnigen Ergebnis führen sollte, wenn es sich vermeiden lässt und sei sie selber noch so fragwürdig.
Es gibt doch gar keinen Grund, das hier so falsch zu implementieren. Weder ist der Code einfacher, noch ist es irgendwie schneller.

Reply:
> Ich gebe nur bei der großen, eine Kapazität von 2 an, bei den beiden anderen nichts, da ich die unverändert lassen möchte. Für die Berechnung des Geasmt-SOC ist ja nur die Relation der Batteriegrößen zueinander wichtig.

Mag sein, wird aber so nicht unterstützt (und ist auch nirgendwo dokumentiert). Dann musst du bei den beiden anderen eben auch eine 1 angeben. Der Aufwand ist beherrschbar.